src/nodes.coffee

```diff
@@ -1610,70 +1611,129 @@ exports.Code = class Code extends Base
 
   # Compilation creates a new scope unless explicitly asked to share with the
-  # outer scope. Handles splat parameters in the parameter list by peeking at
-  # the JavaScript `arguments` object. If the function is bound with the `=>`
-  # arrow, generates a wrapper that saves the current value of `this` through
-  # a closure.
+  # outer scope. Handles splat parameters in the parameter list by setting
+  # such parameters to be the final parameter in the function definition, as
+  # required per the ES2015 spec. If the CoffeeScript function definition had
+  # parameters after the splat, they are declared via expressions in the
+  # function body.
   compileNode: (o) ->
-
-    if @bound and o.scope.method?.bound
-      @context = o.scope.method.context
-
-    # Handle bound functions early.
-    if @bound and not @context
-      @context = '_this'
-      wrapper = new Code [new Param new IdentifierLiteral @context], new Block [this]
-      boundfunc = new Call(wrapper, [new ThisLiteral])
-      boundfunc.updateLocationDataIfMissing @locationData
-      return boundfunc.compileNode(o)
+    if @bound
+      @context = o.scope.method.context if o.scope.method?.bound
+      @context = 'this' unless @context
 
     o.scope         = del(o, 'classScope') or @makeScope o.scope
     o.scope.shared  = del(o, 'sharedScope')
     o.indent        += TAB
     delete o.bare
     delete o.isExistentialEquals
-    params = []
-    exprs  = []
-    for param in @params when param not instanceof Expansion
-      o.scope.parameter param.asReference o
-    for param in @params when param.splat or param instanceof Expansion
-      for p in @params when p not instanceof Expansion and p.name.value
-        o.scope.add p.name.value, 'var', yes
-      splats = new Assign new Value(new Arr(p.asReference o for p in @params)),
-                          new Value new IdentifierLiteral 'arguments'
-      break
-    for param in @params
-      if param.isComplex()
-        val = ref = param.asReference o
-        val = new Op '?', ref, param.value if param.value
-        exprs.push new Assign new Value(param.name), val, '=', param: yes
+    params           = []
+    exprs            = []
+    haveSplatParam   = no
+    paramsAfterSplat = 0
+
+    # Parse the parameters, adding them to the list of parameters to put in the
+    # function definition; and dealing with splats or expansions, including
+    # adding expressions to the function body to declare all parameter
+    # variables that would have been after the splat/expansion parameter.
+    for param, i in @params
+      # Was `...` used with this parameter? (Only one such parameter is allowed per function.)
+      # Splat/expansion parameters cannot have default values, so we need not worry about that.
+      if param.splat or param instanceof Expansion
+        haveSplatParam = yes
+        if param.name.value?
+          splatParamName = param.name.value
```

Comment:
How do I safely get the name of the splat parameter in all cases? Including `@splat...`. I tried `param.asReference` but that triggers a `var` in the function body.

And I have no idea how to handle `[]...`.